src/crimson/osd/pg.cc

```diff
@@ -547,6 +563,10 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
   ObjectContextRef obc,
   const OpInfo &op_info)
 {
+  if (__builtin_expect(stopping, false)) {
```

Comment:
@athanatos @rzarzynski @ronen-fr  could you also take a look at this design? i think @xxhdx1985126 is using `stopping` flag to prevent all continuation from moving forward if a component is being stopped.

probably another approach is to guard all entries offering services with a gate?

Reply:
I actually like the 'flag' approach here: the change here affects many source files, and adding
the flag is a pretty minimal and obvious change that does not interact with the specifics of each service.
@tchaikov : as you suggested: we may want to discuss this during our meeting.

Reply:
@tchaikov I think using a gate has two purposes: 1) prevent further executions; 2) track all ongoing executions. Since all we need is just preventing further executions and a gate may involve more cpu overhead, so I chose the flag way. Am I right?:-)

Reply:
My understanding of the current graceful shutdown procedure:
1. msgr pre-shutdown: stop dispatching events
2. osd components shutdown:
  2.1. stop by preventing new continuations with the "flag";
  2.2. guard possible internal continuations unrelated to event dispatching;
3. msgr shutdown:
  3.1. close all the existing connections;
  3.2. at the same time, prevent new connections (either by accept or connect);
  3.3. wait for related event dispatching continuations finished (external);
  3.4. wait for connection protocol related continuations finished (internal);
4. destruct all;

For historical reasons, messenger not only dispatches events, but also helps to guard event consumption in the dependent components, which I think is beyond messenger's responsibility and result in the complicated procedure above. But to fix this requires more efforts, including:
* make event dispatching to return non-future value unless the connection needs to be throttled (e.g. `ms_dispatch()`);
* guard all entries including event consumption by the according components;

Then I think the shutdown procedure can be simplified:
1. osd components shutdown:
  1.1. stop by preventing new continuations, including event consumption (or simply unregister itself from the messenger dispatcher list);
  1.2. guard all the internal continuations, including event consumption continuations;
2. msgr shutdown:
  2.1. close all the existing connections;
  2.2. at the same time, prevent new connections (either by accept or connect);
  2.3. wait for connection protocol related continuations finished;
3. destruct all;

Reply:
yeah, the proposed shutdown procedure looks elegant, and more *correct* than the existing one.